File: go/tests/parenting_board_test.go
// parenting_board_test.go
//
// Verifies the CircleAI.Parenting port (parenting_board.go): child add/get + name
// ordering, milestone record (blank-ChildId error) + MilestonesFor (newest-first,
// empty for unknown), routine set/get (weekday key, defensive Entries copy), and
// AgeAsOf (at - DOB; unknown-child error).

package circleai_test

import (
	"testing"
	"time"

	circleai "github.com/bhengubv/CircleAI/go"
)

func TestParenting_ChildrenAndMilestones(t *testing.T) {
	b := circleai.NewInMemoryParentingBoard()
	dob := time.Date(2018, 3, 15, 0, 0, 0, 0, time.UTC)
	b.AddChild(circleai.Child{ChildId: "c1", Name: "Sipho", DateOfBirth: dob})
	b.AddChild(circleai.Child{ChildId: "c2", Name: "Ayanda", DateOfBirth: dob})

	if c, ok := b.GetChild("c1"); !ok || c.Name != "Sipho" {
		t.Fatalf("get child = %+v ok=%v", c, ok)
	}
	kids := b.Children()
	if len(kids) != 2 || kids[0].Name != "Ayanda" || kids[1].Name != "Sipho" {
		t.Fatalf("children name order wrong: %+v", kids)
	}

	t0 := time.Date(2026, 7, 1, 0, 0, 0, 0, time.UTC)
	_ = b.RecordMilestone(circleai.Milestone{MilestoneId: "m1", ChildId: "c1", Category: "Motor", Description: "First steps", AchievedAtUtc: t0})
	_ = b.RecordMilestone(circleai.Milestone{MilestoneId: "m2", ChildId: "c1", Category: "Language", Description: "First word", AchievedAtUtc: t0.Add(24 * time.Hour)})
	if err := b.RecordMilestone(circleai.Milestone{MilestoneId: "bad", ChildId: " "}); err == nil {
		t.Fatalf("blank ChildId must error")
	}

	ms := b.MilestonesFor("c1")
	// Newest first: m2 then m1.
	if len(ms) != 2 || ms[0].MilestoneId != "m2" || ms[1].MilestoneId != "m1" {
		t.Fatalf("milestones order wrong: %+v", ms)
	}
	if unknown := b.MilestonesFor("ghost"); len(unknown) != 0 {
		t.Fatalf("unknown child milestones must be empty")
	}
}

func TestParenting_RoutinesAndAge(t *testing.T) {
	b := circleai.NewInMemoryParentingBoard()
	dob := time.Date(2020, 1, 1, 0, 0, 0, 0, time.UTC)
	b.AddChild(circleai.Child{ChildId: "c1", Name: "Lerato", DateOfBirth: dob})

	entries := []circleai.RoutineEntry{{Time: "07:00", Activity: "Wake"}, {Time: "08:00", Activity: "School"}}
	b.SetRoutine(circleai.Routine{ChildId: "c1", DayOfWeek: time.Monday, Entries: entries})
	entries[0].Activity = "MUTATED" // must not affect stored routine

	r, ok := b.GetRoutine("c1", time.Monday)
	if !ok || len(r.Entries) != 2 || r.Entries[0].Activity != "Wake" {
		t.Fatalf("get routine wrong (defensive copy?): %+v ok=%v", r, ok)
	}
	// Different weekday -> absent.
	if _, ok := b.GetRoutine("c1", time.Tuesday); ok {
		t.Fatalf("tuesday routine must be absent")
	}

	at := time.Date(2021, 1, 1, 0, 0, 0, 0, time.UTC)
	age, err := b.AgeAsOf("c1", at)
	if err != nil {
		t.Fatalf("age: %v", err)
	}
	if age != at.Sub(dob) {
		t.Fatalf("age = %v, want %v", age, at.Sub(dob))
	}
	if _, err := b.AgeAsOf("ghost", at); err == nil {
		t.Fatalf("unknown child age must error")
	}
}

File: typescript/tests/parenting_board.test.ts
// parenting_board.test.ts
// Verifies the CircleAI.Parenting port: children (name-ordered), milestones
// (achieved-desc), per-day routines keyed by DayOfWeek name, and age calc.

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import {
  InMemoryParentingBoard,
  ParentingDomainContext,
  DayOfWeek,
  child,
  milestone,
  routine,
  routineEntry,
} from "../src/parenting/index";

const D = (s: string) => new Date(s);

describe("DayOfWeek enum", () => {
  it("matches C# System.DayOfWeek values (Sunday = 0)", () => {
    assert.equal(DayOfWeek.Sunday, 0);
    assert.equal(DayOfWeek.Monday, 1);
    assert.equal(DayOfWeek.Saturday, 6);
  });
});

describe("InMemoryParentingBoard", () => {
  it("adds children ordered by name", () => {
    const b = new InMemoryParentingBoard();
    b.addChild(child("c2", "Zara", D("2018-01-01"), "F"));
    b.addChild(child("c1", "Ali", D("2020-06-15"), "M"));
    assert.deepEqual(
      b.children.map((c) => c.name),
      ["Ali", "Zara"],
    );
    assert.equal(b.getChild("c1")?.gender, "M");
    assert.equal(b.getChild("nope"), undefined);
  });

  it("records milestones newest-first; blank childId throws; empty is []", () => {
    const b = new InMemoryParentingBoard();
    assert.deepEqual(b.milestonesFor("c1"), []);
    b.recordMilestone(milestone("m1", "c1", "Motor", "First steps", D("2021-06-01T00:00:00Z")));
    b.recordMilestone(milestone("m2", "c1", "Speech", "First word", D("2021-09-01T00:00:00Z")));
    assert.deepEqual(
      b.milestonesFor("c1").map((m) => m.milestoneId),
      ["m2", "m1"],
    );
    assert.throws(() => b.recordMilestone(milestone("m", "", "c", "d", D("2021-01-01T00:00:00Z"))), /ChildId required/);
  });

  it("stores and retrieves routines keyed by child + day", () => {
    const b = new InMemoryParentingBoard();
    const mon = routine("c1", DayOfWeek.Monday, [routineEntry("07:00", "Wake"), routineEntry("08:00", "School")]);
    b.setRoutine(mon);
    assert.deepEqual(b.getRoutine("c1", DayOfWeek.Monday)?.entries.length, 2);
    assert.equal(b.getRoutine("c1", DayOfWeek.Tuesday), undefined);
  });

  it("computes age as a duration in ms; unknown child throws", () => {
    const b = new InMemoryParentingBoard();
    b.addChild(child("c1", "Ali", D("2020-01-01T00:00:00Z"), "M"));
    const oneYearMs = D("2021-01-01T00:00:00Z").getTime() - D("2020-01-01T00:00:00Z").getTime();
    assert.equal(b.ageAsOf("c1", D("2021-01-01T00:00:00Z")), oneYearMs);
    assert.throws(() => b.ageAsOf("ghost", D("2021-01-01T00:00:00Z")), /Unknown child ghost/);
  });

  it("domain context exposes prompt + compliance + tools", () => {
    assert.ok(ParentingDomainContext.systemPromptSnippet.includes("[DOMAIN: Parenting]"));
    assert.deepEqual(ParentingDomainContext.complianceFlags, ["Childrens_Act_38_2005", "POPIA"]);
    assert.deepEqual(ParentingDomainContext.suggestedTools, ["development_tracker", "document_editor", "web_search", "calendar"]);
  });
});
